Show the HTTPSOnly browsing warning

sysrqb opened the pull request:
#### 81cb1ae044766532edabd686893e1e4fa415b904
<pre>
Show the HTTPSOnly browsing warning
<a href="https://bugs.webkit.org/show_bug.cgi?id=277554">https://bugs.webkit.org/show_bug.cgi?id=277554</a>
<a href="https://rdar.apple.com/133067048">rdar://133067048</a>

Reviewed by Chris Dumez.

This change shows the browsing warning when a provisional navigation fails and
HTTPSOnly is active. Calling the delegates is now moved into a lambda so that
it can be easily shared with the warning&apos;s completion handler. This is helpful
for both testing and so the client can show more information, if appropriate,
when we call the delegate.

The new browsing warning contains a &quot;Go Back&quot; button and a &quot;Continue&quot; button,
but neither of them perform a specific action. The result of click either
button is calling the delegates, which is the existing behavior. I&apos;ll adjust
this behavior in a follow up patch.

I added and adjusted some API tests for this new warning.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(checkTitleAndClick):
(TEST(WKNavigation, HTTPSOnlyHTTPFallbackGoBack)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallbackContinue)):
(TEST(WKNavigation, HTTPSOnlyWithSameSiteBypass)):
(TEST(WKNavigation, HTTPSOnlyWithHTTPRedirect)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallback)): Deleted.

Canonical link: <a href="https://commits.webkit.org/281914@main">https://commits.webkit.org/281914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5d48f9175ec458daa4661b56eabebcfb643d66ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61423 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40780 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14004 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65374 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11968 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63552 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12243 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49610 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/8312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64470 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37907 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53200 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/30453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34568 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10429 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10881 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56376 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/10728 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67103 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5366 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10502 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5389 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53165 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/57202 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4432 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9239 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36584 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/37667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/38761 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/37412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6319,12 +6320,49 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
+                m_navigationClient->didFailProvisionalLoadWithErrorForFrame(*this, WTFMove(request), error, WTFMove(frameInfo));
+            }
+        };
+#if HAVE(SAFE_BROWSING)
```
I'll rename this macro in another PR.

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6319,12 +6319,36 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
-            m_navigationClient->didFailProvisionalNavigationWithError(*this, FrameInfoData { frameInfo }, navigation.get(), request.url(), error, process->transformHandlesToObjects(userData.protectedObject().get()).get());
-            m_navigationClient->didFailProvisionalLoadWithErrorForFrame(*this, WTFMove(request), error, WTFMove(frameInfo));
-        }
+        auto callClientFunctions = [this, frame = Ref { frame }, navigation, error, process = process, request = WTFMove(request), frameInfo = WTFMove(frameInfo), protectedObject = userData.protectedObject()]() mutable {
```
Minor QoL improvement since we build against C++23 now!
```suggestion
        auto callClientFunctions = [this, frame = Ref { frame }, navigation, error, process = process, request = WTFMove(request), frameInfo = WTFMove(frameInfo), protectedObject = userData.protectedObject()] mutable {
```

nt1m replied:
Not all ports I believe

aprotyas commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6319,12 +6319,36 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
+            if (m_loaderClient)
+                m_loaderClient->didFailProvisionalLoadWithErrorForFrame(*this, frame, navigation.get(), error, process->transformHandlesToObjects(protectedObject.get()).get());
+            else {
+                m_navigationClient->didFailProvisionalNavigationWithError(*this, FrameInfoData { frameInfo }, navigation.get(), request.url(), error, process->transformHandlesToObjects(protectedObject.get()).get());
```
Right?
```suggestion
                m_navigationClient->didFailProvisionalNavigationWithError(*this, WTFMove(frameInfo), navigation.get(), request.url(), error, process->transformHandlesToObjects(protectedObject.get()).get());
```

sysrqb replied:
Can't, because the line after this one moves it

```
m_navigationClient->didFailProvisionalNavigationWithError(*this, FrameInfoData { frameInfo }, navigation.get(), request.url(), error, process->transformHandlesToObjects(protectedObject.get()).get());
m_navigationClient->didFailProvisionalLoadWithErrorForFrame(*this, WTFMove(request), error, WTFMove(frameInfo)); 
```

aprotyas reviewed (commented):
Couple of minor comments. This looks good but I'll leave the r+ to an area expert

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6319,12 +6319,36 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
-            m_navigationClient->didFailProvisionalNavigationWithError(*this, FrameInfoData { frameInfo }, navigation.get(), request.url(), error, process->transformHandlesToObjects(userData.protectedObject().get()).get());
-            m_navigationClient->didFailProvisionalLoadWithErrorForFrame(*this, WTFMove(request), error, WTFMove(frameInfo));
-        }
+        auto callClientFunctions = [this, frame = Ref { frame }, navigation, error, process = process, request = WTFMove(request), frameInfo = WTFMove(frameInfo), protectedObject = userData.protectedObject()]() mutable {
```
nit: I don't think you need the `= process` here.

cdumez replied:
Also, the lambda seems to fail to protect `this`. I think we need to capture a `protectedThis`

sysrqb replied:
> the lambda seems to fail to protect this. I think we need to capture a protectedThis

The lambda is either called synchronously, or we protect `this` in the other completion handler we pass into `protectedPageClient->showBrowsingWarning` (below), but I'll protect `this` in this lambda, so it is clear to anyone else looking at it.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6319,12 +6319,36 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
+            && canFallbackToHTTP) {
+            protectedPageClient->clearBrowsingWarning();
+
+            auto httpFallbackBrowsingWarning = BrowsingWarning::create(failedURL, frame.isMainFrame(), BrowsingWarning::HTTPSNavigationFailureData { });
```
nit: please use an explicit smart pointer type instead of auto.

Commit:
Show the HTTPSOnly browsing warning
https://bugs.webkit.org/show_bug.cgi?id=277554
rdar://133067048

Reviewed by Chris Dumez.

This change shows the browsing warning when a provisional navigation fails and
HTTPSOnly is active. Calling the delegates is now moved into a lambda so that
it can be easily shared with the warning's completion handler. This is helpful
for both testing and so the client can show more information, if appropriate,
when we call the delegate.

The new browsing warning contains a "Go Back" button and a "Continue" button,
but neither of them perform a specific action. The result of click either
button is calling the delegates, which is the existing behavior. I'll adjust
this behavior in a follow up patch.

I added and adjusted some API tests for this new warning.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(checkTitleAndClick):
(TEST(WKNavigation, HTTPSOnlyHTTPFallbackGoBack)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallbackContinue)):
(TEST(WKNavigation, HTTPSOnlyWithSameSiteBypass)):
(TEST(WKNavigation, HTTPSOnlyWithHTTPRedirect)):
(TEST(WKNavigation, HTTPSOnlyHTTPFallback)): Deleted.

Canonical link: https://commits.webkit.org/281914@main
